### PR TITLE
Optimize pipeline of pytorch engine

### PR DESCRIPTION
**Pull request by grimoire**
Less schedule would leads to better gpu utility

- [x] fix window attention

### CODE REVIEW

File: lmdeploy/pytorch/config.py [hunk 1]
```diff
@@ -12,6 +12,20 @@ def _get_torch_dtype(config: Any, default: str = 'float16'):
         config: Config of the hf model.
         default (str): default device type.
     """
+
+    def __hack_qwen():
+        """hack qwen."""
+        torch_dtype = 'bfloat16' if torch.cuda.is_bf16_supported(
```
Comment (lvhan028) [inline lmdeploy/pytorch/config.py:18]: 这个逻辑是不是可以用到所有的模型上？

**Comment by lvhan028** [comment]
- [ ] evaluation test
- [ ] pipeline test
- [ ] cli test
- [ ] api_server test

File: lmdeploy/pytorch/config.py (hunk 1 above)
Comment (grimoire) [reply lmdeploy/pytorch/config.py:18]: 只有 qwen 有这种 key

**Comment by grimoire** [comment]
window attention failed.

File: lmdeploy/pytorch/config.py [hunk 2]
```diff
@@ -12,6 +12,20 @@ def _get_torch_dtype(config: Any, default: str = 'float16'):
         config: Config of the hf model.
         default (str): default device type.
     """
+
```
Comment (RunningLeon) [inline lmdeploy/pytorch/config.py:15]: may fix conflicts with main branch.

File: lmdeploy/pytorch/engine/engine.py [hunk 3]
```diff
@@ -634,564 +498,343 @@ async def __forward(inputs):
...
-
-    async def async_stream_infer(self,
-                                 session_id: int,
-                                 input_ids: List[int],
-                                 gen_config: EngineGenerationConfig = None,
-                                 adapter_name: str = None,
-                                 **kwargs):
-        """Send stream inference request.
+    def create_instance(self, cuda_stream_id=0):
+        """Create a turbomind instance.
```
Comment (RunningLeon) [inline lmdeploy/pytorch/engine/engine.py:765]: turbomind?

File: lmdeploy/pytorch/engine/engine.py [hunk 4]
```diff
@@ -634,564 +498,343 @@ async def __forward(inputs):
...
-            gen_config (EngineGenerationConfig): The sampling parameters.
-            adapter_name (str): The lora adapter name.
-
-        Yields:
-            int: Error flags. 0 if success.
-            List[int]: The streaming output tokens.
-            int: The number of the output tokens.
+            cuda_stream_id(int): identity of a cuda stream
+        Returns:
+            EngineInstance: an instance of turbomind
```
Comment (RunningLeon) [inline lmdeploy/pytorch/engine/engine.py:770]: turbomind -> pytorch engine?

**Comment by zhoujh1113** [comment]
长文本问题回答出很多'\n'：
img_v3_029h_df0911ee-e311-4aec-8543-8f6c94bd3bfg
img_v3_029h_ffa99caf-c8d1-4b2e-8231-293a7f45299g

**Comment by grimoire** [comment]
@zhoujh1113 Pull the latest commit and try again

File: lmdeploy/pytorch/engine/logits_process.py [hunk 5]
```diff
@@ -144,8 +148,13 @@ def __get_topp(top_p):
...
             return top_p, min_top_p
 
-        def __get_bad_words(bad_words, max_bw_len):
+        def __get_bad_words(bad_words):
             """get bad words."""
+            max_bw_len = max(len(bw) for bw in bad_words)
+            if max_bw_len == 0:
+                return None
+            if all(len(bw) == max_bw_len for bw in bad_words):
+                return torch.tensor(bad_words)
```
Comment (RunningLeon) [inline lmdeploy/pytorch/engine/logits_process.py:157]: fails with Qwen-14B-Chat with pipeline
```
Token indices sequence length is longer than the specified maximum sequence length for this model (8371 > 8192). Running this sequence through the model will result in indexing errors
2024-04-01 16:36:39,430 - lmdeploy - ERROR - Engine loop failed with error: Could not infer dtype of NoneType
Traceback (most recent call last):
  File "/home/maningsheng/projects/lmdeploy/lmdeploy/pytorch/engine/request.py", line 17, in _raise_exception_on_finish
    task.result()
  File "/home/maningsheng/projects/lmdeploy/lmdeploy/pytorch/engine/engine.py", line 756, in async_loop
    await __step(True)
  File "/home/maningsheng/projects/lmdeploy/lmdeploy/pytorch/engine/engine.py", line 742, in __step
    raise e
  File "/home/maningsheng/projects/lmdeploy/lmdeploy/pytorch/engine/engine.py", line 734, in __step
    raise out
  File "/home/maningsheng/projects/lmdeploy/lmdeploy/pytorch/engine/engine.py", line 675, in _async_loop_background
    sampling_inputs = SamplingInputs.from_sampling_params(running)
  File "/home/maningsheng/projects/lmdeploy/lmdeploy/pytorch/engine/logits_process.py", line 177, in from_sampling_params
    stop_words = __get_bad_words(stop_words)
  File "/home/maningsheng/projects/lmdeploy/lmdeploy/pytorch/engine/logits_process.py", line 157, in __get_bad_words
    return torch.tensor(bad_words)
RuntimeError: Could not infer dtype of NoneType
```

**Comment by zhoujh1113** [comment]
> @zhoujh1113 Pull the latest commit and try again

试过可以了
图片

File: lmdeploy/pytorch/engine/logits_process.py (hunk 5 above)
Comment (grimoire) [reply lmdeploy/pytorch/engine/logits_process.py:157]: Fixed

**Comment by zhulinJulia24** [comment]
https://github.com/InternLM/lmdeploy/actions/runs/8506126390/job/23297552003   发起了pytorch case验证，覆盖pytorch chat、restful和pipeline
功能两个case失败，一个手动复现通过。另外一个是已知deepseek-moe回答效果较差的问题

image



restful接口测试 @grimoire 
发现stream下output输出变了？和turbomind一致了？

    def test_chat_interactive_max_tokens_stream(self):
        api_client = APIClient(BASE_URL)
        outputList = []
        for output in api_client.chat_interactive_v1(
                prompt='Hi, pls intro yourself',
                stream=True,
                request_output_len=5,
                temperature=0.01):
            outputList.append(output)
        assert_chat_interactive_stream_return(outputList[-1],
                                              True,
                                              index=len(outputList) - 2)
        for index in range(0, len(outputList) - 1):
            assert_chat_interactive_stream_return(outputList[index],
                                                  index=index)
        assert output.get('finish_reason') == 'length'
>       assert len(outputList) == 6
E       AssertionError: assert 7 == 6

**Review by zhulinJulia24 — APPROVED** [review]
lgtm

**Review by RunningLeon — APPROVED** [review]
LGTM

**Pull request merged**